File: P5 2024-2/Paradigmas de Programacao/ListaExercicio01/Golang/Empresa.go
package main

import "fmt"

type Funcionario struct { // será o funcionario assalariado para questao 11
	nome    string
	cargo   string
	salario float64
}

type Empresa struct {
	nome         string
	Funcionarios []*Funcionario
}

func (e *Empresa) contrataFuncionario(f *Funcionario) {
	e.Funcionarios = append(e.Funcionarios, f)
}

func (e *Empresa) listaFuncionarios() {
	fmt.Println("Listando funcionarios da empresa ", e.nome)
	for _, f := range e.Funcionarios {
		fmt.Println(f.nome)
	}
}

type calcularSalario interface {
	calcularSalario()
}

func (f *Funcionario) calcularSalario() float64 {
	fmt.Println("Salário de", f.nome, ": R$ ", f.salario)
	return f.salario
}

type FuncionarioHorista struct {
	nome         string
	cargo        string
	precoHora    float64
	cargaHoraria int
}

func (fh *FuncionarioHorista) calcularSalario() float64 {
	fmt.Println("Salário de", fh.nome, ": R$ ", (fh.precoHora * float64(fh.cargaHoraria)))
	return (fh.precoHora * float64(fh.cargaHoraria))
}

File: P5 2024-2/Paradigmas de Programacao/ListaExercicio01/Python/Empresa.py
class abstractCalcularSalario:
    def calcularSalario(self):
        pass

class Empregado:
    def __init__(self, nome, cargo, salario):
        self.nome = nome
        self.cargo = cargo
        self.salario = salario

    def exibeDetalhes(self):
        print(f"Nome: {self.nome}, Cargo: {self.cargo}, Salario: R$ {self.salario}")

    def calcularSalario(self):
        print(f"Salário de {self.nome}: R$ {self.salario}")
        return self.salario

class EmpregadoHorista:
    def __init__(self, nome, cargo, valor_hora, horas):
        self.nome = nome
        self.cargo = cargo
        self.valor_hora = valor_hora
        self.horas = horas

    def exibeDetalhes(self):
        print(f"Nome: {self.nome}, Cargo: {self.cargo}")

    def calcularSalario(self):
        salario = self.valor_hora * self.horas
        print(f"Salário de {self.nome}: R$ {salario}")
        return salario


class Empresa:
    def __init__(self, nome):
        self.nome = nome
        self.Empregados = []

    def contrataEmpregado(self, Empregado):
        self.Empregados.append(Empregado)

    def listaEmpregados(self):
        print("Listando Empregados da empresa", self.nome)
        for p in self.Empregados:
            p.exibeDetalhes()